File: website/vo2-go/timestamp.go
package vo2

import (
	"strconv"
	"strings"
)

func isLeap(year int64) bool {
	return (year%4 == 0 && year%100 != 0) || year%400 == 0
}

// parseTimestamp parses ISO 8601 timestamps like "2024-01-15T10:30:00Z"
// or "2024-01-15T10:30:00.000Z" into seconds since 2000-01-01.
// Returns ok=false on any parse or range error.
func parseTimestamp(s string) (float64, bool) {
	s = strings.TrimSpace(s)
	s = strings.TrimSuffix(s, "Z")
	dateStr, timeStr, ok := strings.Cut(s, "T")
	if !ok {
		return 0, false
	}

	dp := strings.Split(dateStr, "-")
	if len(dp) != 3 {
		return 0, false
	}
	year, err1 := strconv.ParseInt(dp[0], 10, 64)
	month, err2 := strconv.ParseInt(dp[1], 10, 64)
	day, err3 := strconv.ParseInt(dp[2], 10, 64)
	if err1 != nil || err2 != nil || err3 != nil {
		return 0, false
	}

	// Drop fractional seconds
	if dot := strings.IndexByte(timeStr, '.'); dot >= 0 {
		timeStr = timeStr[:dot]
	}
	tp := strings.Split(timeStr, ":")
	if len(tp) != 3 {
		return 0, false
	}
	hour, err4 := strconv.ParseFloat(tp[0], 64)
	min, err5 := strconv.ParseFloat(tp[1], 64)
	sec, err6 := strconv.ParseFloat(tp[2], 64)
	if err4 != nil || err5 != nil || err6 != nil {
		return 0, false
	}

	if year < 2000 || year > 2100 {
		return 0, false
	}
	if month < 1 || month > 12 {
		return 0, false
	}
	var maxDay int64
	switch month {
	case 2:
		if isLeap(year) {
			maxDay = 29
		} else {
			maxDay = 28
		}
	case 4, 6, 9, 11:
		maxDay = 30
	default:
		maxDay = 31
	}
	if day < 1 || day > maxDay {
		return 0, false
	}
	if hour < 0 || hour > 23 || min < 0 || min > 59 || sec < 0 || sec > 59 {
		return 0, false
	}

	var days int64
	for y := int64(2000); y < year; y++ {
		if isLeap(y) {
			days += 366
		} else {
			days += 365
		}
	}
	monthDays := [12]int64{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}
	for m := int64(0); m < month-1; m++ {
		days += monthDays[m]
		if m == 1 && isLeap(year) {
			days++
		}
	}
	days += day - 1

	return float64(days)*86400.0 + hour*3600.0 + min*60.0 + sec, true
}

File: website/vo2-cs/Timestamp.cs
using System.Globalization;

namespace Vo2;

internal static class Timestamp
{
    private static bool IsLeap(long year) => (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;

    /// <summary>
    /// Parses an ISO 8601 timestamp like "2024-01-15T10:30:00Z" or
    /// "2024-01-15T10:30:00.000Z" into seconds since 2000-01-01.
    /// </summary>
    public static bool TryParse(string s, out double seconds)
    {
        seconds = 0;
        s = s.Trim();
        if (s.EndsWith('Z')) s = s[..^1];

        var tIdx = s.IndexOf('T');
        if (tIdx < 0) return false;
        var dateStr = s[..tIdx];
        var timeStr = s[(tIdx + 1)..];

        var dp = dateStr.Split('-');
        if (dp.Length != 3) return false;
        if (!long.TryParse(dp[0], NumberStyles.Integer, CultureInfo.InvariantCulture, out var year)
            || !long.TryParse(dp[1], NumberStyles.Integer, CultureInfo.InvariantCulture, out var month)
            || !long.TryParse(dp[2], NumberStyles.Integer, CultureInfo.InvariantCulture, out var day))
        {
            return false;
        }

        var dotIdx = timeStr.IndexOf('.');
        if (dotIdx >= 0) timeStr = timeStr[..dotIdx];

        var tp = timeStr.Split(':');
        if (tp.Length != 3) return false;
        if (!double.TryParse(tp[0], NumberStyles.Float, CultureInfo.InvariantCulture, out var hour)
            || !double.TryParse(tp[1], NumberStyles.Float, CultureInfo.InvariantCulture, out var min)
            || !double.TryParse(tp[2], NumberStyles.Float, CultureInfo.InvariantCulture, out var sec))
        {
            return false;
        }

        if (year < 2000 || year > 2100) return false;
        if (month < 1 || month > 12) return false;
        long maxDay = month switch
        {
            2 => IsLeap(year) ? 29 : 28,
            4 or 6 or 9 or 11 => 30,
            _ => 31,
        };
        if (day < 1 || day > maxDay) return false;
        if (hour < 0 || hour > 23 || min < 0 || min > 59 || sec < 0 || sec > 59) return false;

        long days = 0;
        for (long y = 2000; y < year; y++) days += IsLeap(y) ? 366 : 365;
        ReadOnlySpan<int> monthDays = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
        for (int m = 0; m < month - 1; m++)
        {
            days += monthDays[m];
            if (m == 1 && IsLeap(year)) days++;
        }
        days += day - 1;

        seconds = days * 86400.0 + hour * 3600.0 + min * 60.0 + sec;
        return true;
    }
}
